File: Override02.rb
# 계산기 예제 override 적용해보기

class Cal
    @@_history = []  
    def initialize(v1,v2) 
        @v1 = v1
        @v2 = v2          
    end
    def add()
        result = @v1+@v2 
        @@_history.push("add : #{@v1}+#{@v2}=#{result}") 
        return result
    end
    def subtract()
        result = @v1-@v2 
        @@_history.push("subtract : #{@v1}-#{@v2}=#{result}")
        return result
    end
    def Cal.history() 
        for item in @@_history 
            p item
        end
    end
    def info()
        return "Cal : v1 = #{@v1} , v2 = #{@v2}"
    end
end

class CalMultiply < Cal
    def multiply()
        result = @v1*@v2 
        @@_history.push("multiply : #{@v1}*#{@v2}=#{result}")
        return result
    end
    def info()
        return "CalMultiply => #{super()}"
    end
end

class CalDivide < CalMultiply
    def divide()
        result = @v1/@v2 
        @@_history.push("divide : #{@v1}/#{@v2}=#{result}")
        return result
    end
    def info()
        return "CalDivide => #{super()}"
    end
end


c0 = Cal.new(30, 60)
p c0.info()
c1 = CalMultiply.new(10, 10)
p c1.info()
c2 = CalDivide.new(20, 10)
p c2.info()

File: Inheritance/Override02.py
# 계산기 예제에서 override 활용

class Cal(object):
    _history = []

    def __init__(self, v1, v2):
        self.v1 = v1
        self.v2 = v2

    def add(self):
        result = self.v1+self.v2
        Cal._history.append(
            "add : {0}+{1}={2}".format(self.v1, self.v2, result))
        return result

    def subtract(self):
        result = self.v1-self.v2
        Cal._history.append("subtract : %d-%d=%d" % (self.v1, self.v2, result))
        return result

    @classmethod
    def history(cls):
        for item in Cal._history:
            print(item)

    def info(self):
        return "Cal => v1 : %d, v2 : %d" % (self.v1, self.v2)


class CalMultiply(Cal):
    def multiply(self):
        result = self.v1*self.v2
        Cal._history.append("multiply : %d*%d=%d" % (self.v1, self.v2, result))
        return result

    def info(self):
        return "CalMultiply => %s" % (super().info())


class CalDivide(CalMultiply):
    def divide(self):
        result = self.v1/self.v2
        Cal._history.append("divide : %d/%d=%d" % (self.v1, self.v2, result))
        return result

    def info(self):
        return "CalDivdie => %s" % (super().info())


c0 = Cal(30, 60)
print(c0.info())
c1 = CalMultiply(10, 10)
print(c1.info())
c2 = CalDivide(20, 10)
print(c2.info())
